make dyanmo work with _LazyGraphModule.lazy_forward

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121259

Fix https://github.com/pytorch/pytorch/issues/121198 .

We previously already trigger the real recompilation for LazyGraphModule when it runs thru dynamo context. But people may pass in LazyGraphModule._lazy_forward rather than the LazyGraphModule instance itself. This PR handles that.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

shunting314 committed:
make dyanmo work with _LazyGraphModule.lazy_forward

[ghstack-poisoned]

williamwen42 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -356,21 +356,31 @@ def get_compiler_config():
         fn = innermost_fn(fn)
 
         # add context containing GraphModule to any GraphModule forward functions
-        if isinstance(fn, torch.fx.GraphModule):
```
Why are we removing this case? It seems like `test_preserve_fx_node_metadata` is failing as a result.

shunting314 replied:
Didn't realize some code is ran for non _LazyGraphModule in this code block. Fixing it now.

jansel reviewed (changes requested):
Tests are failing

shunting314 committed:
Update on "make dyanmo work with _LazyGraphModule.lazy_forward"


Fix https://github.com/pytorch/pytorch/issues/121198 .

We previously already trigger the real recompilation for LazyGraphModule when it runs thru dynamo context. But people may pass in LazyGraphModule._lazy_forward rather than the LazyGraphModule instance itself. This PR handles that.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

shunting314 commented:
@pytorchbot merge
